src/core/ipe-setup.c

```diff
@@ -0,0 +1,165 @@
+#include "path-util.h"
+
+#define IPE_SECFS_DIR "/sys/kernel/security/ipe"
+#define IPE_SECFS_NEW_POLICY IPE_SECFS_DIR "/new_policy"
+#define IPE_SECFS_POLICIES IPE_SECFS_DIR "/policies/"
+#define IPE_POLICY_PATH "/etc/ipe/ipe-policy.p7b"
+
+int ipe_setup(void) {
+#if ENABLE_IPE
+        _cleanup_free_ char *policy_name = NULL, *policy_version = NULL, *policy_existing_version = NULL,
+                *policy_existing_version_file = NULL, *policy_activate = NULL, *output_path_free = NULL;
+        _cleanup_fclose_ FILE *finput = NULL;
+        _cleanup_close_ int input = -EBADF, output = -EBADF;
+        const char *output_path;
+        int r;
+
+        if (access(IPE_SECFS_DIR, F_OK) < 0) {
+                log_debug_errno(errno, "IPE support is disabled in the kernel, ignoring: %m");
+                return 0;
+        }
+
+        if (access(IPE_POLICY_PATH, F_OK) < 0) {
+                log_debug_errno(errno, "No IPE custom policy file "IPE_POLICY_PATH", ignoring: %m");
+                return 0;
+        }
+
+        input = open(IPE_POLICY_PATH, O_RDONLY|O_CLOEXEC);
+        if (input < 0)
+                return log_error_errno(
+                                errno,
+                                "Failed to open the IPE custom policy file "IPE_POLICY_PATH", ignoring: %m");
+
+        r = fdopen_independent(input, "re", &finput);
+        if (r < 0)
+                return log_error_errno(
+                                r,
+                                "Failed to fdopen the IPE custom policy file " IPE_POLICY_PATH ", ignoring: %m");
+
+        /* First we need to parse the policy name, since in order to activate it we need to know it, as it
+         * will show up as a directory under /sys/kernel/security/ipe/policies/ and there will be an
+         * 'active' file under it to write into, which will make it active. */
+        for (;;) {
+                _cleanup_free_ char *line = NULL;
+                const char *p;
+
+                r = read_line(finput, LONG_LINE_MAX, &line);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to read the IPE policy "IPE_POLICY_PATH": %m");
+                if (r == 0)
+                        break;
+
+                for (p = line;;) {
+                        _cleanup_free_ char *word = NULL;
+                        const char *key;
+
+                        r = extract_first_word(&p, &word, WHITESPACE, /* flags= */ 0);
+                        if (r < 0)
+                                return log_error_errno(
+                                                r,
+                                                "Failed to extract the first word from line '%s' from IPE policy " IPE_POLICY_PATH
+                                                ": %m",
+                                                line);
+                        if (r == 0)
+                                break;
+
+                        key = strstrafter(word, "policy_name=");
```

Comment:
This can be switched to startswith() now that extract_first_word() takes care of whitespace.

Reply:
It's a file with inline binary signature, so a line can look like this `��R�Npolicy_name=verity policy_version=0.0.1` hence we need to search through it, not just the beginning of it, to skip over the rest of the data

Reply:
hmm, what? what is "an inline binary signature" supposed to be? it's prefixes with pkcs7 or so? uh?

Reply:
p7b format see https://github.com/systemd/systemd/pull/34418#discussion_r1759795995